library/std/src/alloc.rs

```diff
@@ -127,6 +127,115 @@ pub use alloc_crate::alloc::*;
 #[derive(Debug, Default, Copy, Clone)]
 pub struct System;
 
+use crate::sys::alloc::System as Imp;
+
+// When debug assertions are not enabled, `System` just forwards down to the particular platform
+// implementation.
+#[cfg(not(debug_assertions))]
+#[stable(feature = "alloc_system_type", since = "1.28.0")]
+unsafe impl GlobalAlloc for System {
+    #[inline]
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        unsafe { Imp.alloc(layout) }
+    }
+
+    #[inline]
+    unsafe fn alloc_zeroed(&self, layout: Layout) -> *mut u8 {
+        unsafe { Imp.alloc_zeroed(layout) }
+    }
+
+    #[inline]
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
+        unsafe { Imp.dealloc(ptr, layout) }
+    }
+
+    #[inline]
+    unsafe fn realloc(&self, ptr: *mut u8, layout: Layout, new_size: usize) -> *mut u8 {
+        unsafe { Imp.realloc(ptr, layout, new_size) }
+    }
+}
+
+// Some system allocators (most notably any provided by calling malloc) will always return pointers
+// with an alignment of 8. So for any allocation with an alignment less than 8, we increase the
+// alignment to 8 and return a pointer which is offset into the allocation such that it is not
+// over-aligned.
+// We always bump up the size of an allocation by 8 when the alignment is less than 8.
+#[cfg(debug_assertions)]
+trait LayoutExt {
+    fn with_alignment_padding(self) -> Self;
+    unsafe fn add_alignment_padding(self, ptr: *mut u8) -> *mut u8;
+    unsafe fn remove_alignment_padding(self, ptr: *mut u8) -> *mut u8;
+}
+#[cfg(debug_assertions)]
+impl LayoutExt for Layout {
+    fn with_alignment_padding(self) -> Self {
+        if self.align() < 8 {
+            Layout::from_size_align(self.size() + (8 - self.align()), 8).unwrap()
+        } else {
+            self
+        }
+    }
+    unsafe fn add_alignment_padding(self, ptr: *mut u8) -> *mut u8 {
+        if !ptr.is_null() && self.align() < 8 {
+            // SAFETY: This must be called on a pointer previously returned by a padded Layout,
+            // which will always have space to do this offset
+            unsafe { ptr.add(8 - self.align()) }
+        } else {
+            ptr
+        }
+    }
+    unsafe fn remove_alignment_padding(self, ptr: *mut u8) -> *mut u8 {
```

Comment:
Should/can we check to ensure the input pointer has the lowest bits set according to the align, and abort if not (because that means they gave us the wrong alignment, which happens).

Reply:
Are you referring to detecting an attempt to `dealloc` a different pointer value than one returned by `alloc`? I think that's a useful property for allocators to have, but it seems like glibc's allocator already does the full implementation of that (as opposed to just checking a few bits), and perhaps as a result people don't make this mistake.

In this PR I'm more interested to know if t-libs is willing to take such an allocator, and I'm starting with a behavior that I know exposes a lot of latent bugs in the wild. We can always add more things to this later if it is accepted.

Reply:
No, I mean if you alloc with align 2, then try and dealloc with align 1, the offset calculated will be wrong, wouldn't it? Because the offset is calculated based on the alignment.

Recovering the alignment from the given to `dealloc` address (if it ends in `0b111`, assume alignment 1, if it ends in `0b110`, assume 2, `0b100` is 4) and then aborting with an error if it's wrong, because we would have tried to free with the wrong address seems sensible.

As in, the address we were given was the same as the address we passed back from alloc, but the *alignment* is different.

Reply:
Ahhhhh I see. Yes the current implementation is definitely not good here, because it turns code which is UB but works into an error that only makes any sense if you know about this implementation.

I think eventually we should detect and report that error, but for this PR I am wary of scope creep and I'm not even sure how to print in an allocator.